contracts/SafeProtocolMediator.sol

```diff
@@ -0,0 +1,160 @@
+// SPDX-License-Identifier: LGPL-3.0-only
+pragma solidity ^0.8.18;
+import {ISafeProtocolMediator} from "./interfaces/Mediator.sol";
+import {ISafeProtocolModule} from "./interfaces/Components.sol";
+
+import {ISafe} from "./interfaces/Accounts.sol";
+import {SafeProtocolAction, SafeTransaction, SafeRootAccess} from "./DataTypes.sol";
+import {Ownable2Step} from "@openzeppelin/contracts/access/Ownable2Step.sol";
+
+/**
+ * @title TODO
+ * @notice TODO
+ */
+contract SafeProtocolMediator is ISafeProtocolMediator, Ownable2Step {
+    mapping(address => uint) public nonces;
+
+    struct MoudleAccessInfo {
+        bool enabled;
+        bool rootAddressGranted;
+        // TODO: Add deadline for validity
+    }
+
+    /**
+     * @notice Mapping of a mapping what stores information about modules that are enabled per Safe.
+     *         address (Safe address) => address (component address) => EnabledMoudleInfo
+     */
+    mapping(address => mapping(address => MoudleAccessInfo)) public enabledComponents;
+
+    event ActionsExecuted(address safe, bytes32 metaHash);
+    event RootAccessActionsExecuted(address safe, bytes32 metaHash);
+    event ModuleEnabled(address safe, address module, bool allowRootAccess);
+    event ModuleDisabled(address safe, address module);
+
+    error InvalidNonce(address sender, uint256 nonce);
+    error ModuleRequiresRootAccess(address sender);
+    error MoudleNotEnabled(address module);
+    error ModuleEnabledOnlyForRootAccess(address module);
+    error ModuleAccessMismatch(address module, bool requiresRootAccess, bool providedValue);
+
+    constructor(address initalOwner) {
+        _transferOwnership(initalOwner);
+    }
+
+    modifier onlyEnabledModule(ISafe safe) {
+        if (!enabledComponents[address(safe)][msg.sender].enabled) {
+            revert MoudleNotEnabled(msg.sender);
+        }
+        _;
+    }
+
+    /**
+     * @notice TODO
+     * @param safe TODO
+     * @param transaction TODO
+     * @return success TODO
+     * @return data TODO
+     */
+    function executeTransaction(
+        ISafe safe,
+        SafeTransaction calldata transaction
+    ) external override onlyEnabledModule(safe) returns (bool success, bytes[] memory data) {
+        // TODO: Check for re-entrancy attacks
+        // TODO: Validate metahash
+
+        if (ISafeProtocolModule(msg.sender).requiresRootAccess()) {
+            revert ModuleRequiresRootAccess(msg.sender);
+        }
+
+        if (enabledComponents[address(safe)][msg.sender].rootAddressGranted) {
+            revert ModuleEnabledOnlyForRootAccess(msg.sender);
+        }
+
+        if (nonces[msg.sender] != transaction.nonce) {
```

Comment:
I would not actually track nonces. Making sure that transactions are unique is the job of the module

Reply:
I added this to the specs: https://github.com/5afe/safe-protocol-specs/pull/2/files